### DISCUSSION TITLE
Iterative Schwarz method for domain decomposition

### DISCUSSION THREAD

**Discussion by salaudeen-ya** [idaholab/moose#22516] [Q&A Modules: General] [answered]
Hello,

I have a 1D domain to be decomposed into two subdomains by solving the Poisson equation (at least) using the Additive Schwarz Algorithm.

My enquiries:
+ Can this be solved with MOOSE?
+ How do I go about it? (The modules, multiapp solver, transfers etc).

I have a little understanding of MOOSE as I have been able to use it solve some IB basic FEM analysis of Diffusion-Reaction-Source equation.

I would be glad to have as much as as possible.

Thank you!

----The Equation---
![image](https://user-images.githubusercontent.com/38023107/198358694-7d14d658-9318-4a33-92e5-47db383bfc79.png)

---The Algorithm-------
![image](https://user-images.githubusercontent.com/38023107/198358538-e419d886-304c-463d-9483-73f9de003c1d.png)

**Comment by GiudGiud** [comment, marked as answer]
Hello

We support using asm preconditioning from petsc if you want to try that with a simple Poisson problem
There are examples for poisson in the tutorial and the petsc options are set in the executioner block
https://mooseframework.inl.gov/source/executioners/Steady.html

Guillaume

**Reply by GiudGiud** [reply]
otherwise if you want to write your own custom iterative method, then this is possible in MOOSE as well

**Reply by salaudeen-ya** [reply]
Hello, thank you for your response.

I understand I can use the petsc in the executioner block of the input file. I am actually looking for how to do this with Multi-App System where I have two different systems that run iteratively, and only transfer the solution at the interface. The emphasis is on using Mullti-App  (with different input files) and transfer so as to have the same solution as when the problem is solved with a single input file.

For instance, I would like to solve this problem between [-1,0] using two subdomains [-1,0] and [0,1] with interface at point (0,0). It admits DirichletBC on the left and NeumanBC on the right. The idea is to enforce u and v at the interface, and to make the flux of the two variables to be the negative of the other at the interface. For some reasons, I amnot getting the results as expected and my first domain (sub1) is returning zero values for the flux and u. So far, I have the following trials:
```
---- Sub 1 ---------
[Mesh]
  type = GeneratedMesh
  dim = 1
  nx = 64
  xmin = -1
  xmax = 0
[]

[Variables]
  [u]
    order = FIRST
    family = LAGRANGE
  []
[]

[AuxVariables]
  [flux_u]
    order = CONSTANT
    family = MONOMIAL
  
  []
[]

[Kernels]
  [diffusion]
    type = MatDiffusion
    variable = u
    diffusivity = 2
  []
[]

[AuxKernels]
  [fluxKernel]
    type = DiffusionFluxAux
    diffusivity = 2
    variable = flux_u
    diffusion_variable = u
    component = x
  []
[]

[BCs]
  [leftBC] 
    type = DirichletBC
    variable = u
    boundary = 'left'
    value = 0
  []

  [rightBC_Neumann] 
    type = FunctionNeumannBC
    variable = u
    boundary = 'right' 
    function =  BC_u_right_Neumann
    #preset = true
  []

  [rightBC_Dirichlet] 
    type = FunctionDirichletBC
    variable = u
    boundary = 'right' 
    function = BC_u_right_Dirichlet
    preset = true
  []
[]


[Postprocessors]
  [right_uflux]
    type = PointValue
    variable = 'flux_u'
    point = '0 0 0'
   []
  [right_uval]
    type = PointValue
    variable = 'u'
    point = '0 0 0'
   []
  [vval_left]
    type = Receiver
  []
  [v_flux_left]
    type = Receiver
  []
[]

[Functions]
  [BC_u_right_Neumann]
    type = ParsedFunction
    value = '-a'
    vars = 'a'
    vals = 'v_flux_left'
  []
  [BC_u_right_Dirichlet]
    type = ParsedFunction
    value = 'b'
    vars = 'b'
    vals = 'vval_left'
  []
[]

[Executioner]
  type = Steady
  solve_type = 'PJFNK'

  fixed_point_max_its = 20
  nl_abs_tol = 1e-10
  fixed_point_rel_tol = 1e-16
  fixed_point_abs_tol = 1e-10

  petsc_options_iname = '-pc_type -pc_hypre_type'
  petsc_options_value = 'hypre boomeramg'
[]

[Outputs]
  exodus = true
  #csv = true
  console = true
[]


[MultiApps]
  [sub]
    type = FullSolveMultiApp
    execute_on = 'timestep_begin'
    positions = '0 0 0'
    input_files = sub2.i 
  []
[]


[Transfers]
  [flux_ut]
    type = MultiAppPostprocessorTransfer
    to_multi_app = sub
    from_postprocessor = right_uflux
    to_postprocessor = u_flux_right
  []
  [uval_t]
    type = MultiAppPostprocessorTransfer
    to_multi_app = sub
    from_postprocessor = right_uval
    to_postprocessor = uval_right
  []
  [flux_vt]
    type = MultiAppPostprocessorTransfer
    from_multi_app = sub
    reduction_type = average
    from_postprocessor = left_vflux
    to_postprocessor = v_flux_left
  []
  [vval_t]
    type = MultiAppPostprocessorTransfer
    from_multi_app = sub
    reduction_type = average
    from_postprocessor = left_vval
    to_postprocessor = vval_left
  [] 
[]
```

```
-----Sub 2 -----
[Mesh]
  type = GeneratedMesh
  dim = 1
  nx = 64
  xmin = 0
  xmax = 1

[]

[Variables]
  [v]
    order = FIRST
    family = LAGRANGE
  []
[]

[AuxVariables]
  [flux_v]
    order = CONSTANT
    family = MONOMIAL
  []
[]

[Kernels]
  [diffusion]
    type = MatDiffusion
    variable = v
    diffusivity = 2
  []
[]

[AuxKernels]
  [fluxKernel]
    type = DiffusionFluxAux
    diffusivity = 2
    variable = flux_v
    diffusion_variable = v
    component = x
  []
[]

[BCs]
  [rightBC] 
    type = NeumannBC
    variable = v
    boundary = 'right'
    value = 3
  []

  [leftBC_Neumann] 
    type = FunctionNeumannBC
    variable = v
    boundary = 'left' 
    function = BC_vleft_Neumann
  []

  [leftBC_Dirichlet] 
    type = FunctionDirichletBC
    variable = v
    boundary = 'left' 
    function = BC_vleft_Dirichlet
    preset = true
  []

[]

[Postprocessors]
  [left_vflux]
    type = PointValue
    variable = 'flux_v'
    point = '0 0 0'
   []
  [left_vval]
    type = PointValue
    variable = 'v'
    point = '0 0 0'
   []
  [uval_right]
    type = Receiver
  []
  [u_flux_right]
    type = Receiver
  []
[]

[Functions]
  [BC_vleft_Neumann]
    type = ParsedFunction
    value = '-a'
    vars = 'a'
    vals = 'u_flux_right'

  []
  [BC_vleft_Dirichlet]
    type = ParsedFunction
    value = 'b'
    vars = 'b'
    vals = 'uval_right'
  []
[]


[Executioner]
  type = Steady
  solve_type = 'PJFNK'

  fixed_point_max_its = 20
  nl_abs_tol = 1e-10
  fixed_point_rel_tol = 1e-16
  fixed_point_abs_tol = 1e-10

  petsc_options_iname = '-pc_type -pc_hypre_type'
  petsc_options_value = 'hypre boomeramg'
[]

[Outputs]
  exodus = true
  #csv = true
  console = true
[]
```


```
----- The reference domain (when solved as a single domain) -------
[Mesh]
  type = GeneratedMesh
  dim = 1
  nx = 64
  xmin = -1
  xmax = 1

[]

[Variables]
  [v]
    order = FIRST
    family = LAGRANGE
  []
[]

[AuxVariables]
  [flux_v]
    order = CONSTANT
    family = MONOMIAL
  []
[]

[Kernels]
  [diffusion]
    type = MatDiffusion
    variable = v
    diffusivity = 2
  []
[]

[AuxKernels]
  [fluxKernel]
    type = DiffusionFluxAux
    diffusivity = 2
    variable = flux_v
    diffusion_variable = v
    component = x
  []
[]

[BCs]
  [rightBC] 
    type = NeumannBC
    variable = v
    boundary = 'right'
    value = 3
  []

  [leftBC] 
    type = DirichletBC
    variable = v
    boundary = 'left'
    value = 0
  []

[]


[Executioner]
  type = Steady
  solve_type = 'PJFNK'
[]

[Outputs]
  exodus = true
  #csv = true
  console = true
[]

```

------

Thank you! I hope someone can help spot the problem.

**Reply by GiudGiud** [reply]
so on the right of subapp1 you have a dirichlet and a neumann BC. 
Is that desired? Are you trying to have like a robin BC ?

**Comment by salaudeen-ya** [comment]
Yes. Exactly. And I've tried the VacuumBC too which doesn't give me the
desired results.


How do I go about doing that?

**Reply by GiudGiud** [reply]
I m actually not sure, I just dont think it's with Dirichlet + Neumann that we make those. This should just set the value of the variable, and the flux, as if you had two equations at the boundary. 

We have robinBCs in the electromagnetics module, they are fairly complicated though. We have other RobinBC in the test object but they dont look right

@cticenhour do you know how to make the simple case RobinBC ?

**Reply by cticenhour** [reply]
Yeah, we don't add the impact of Dirichlet + Neumann like we do with kernels. In fact, I believe Dirichlet actually manipulates the matrix by zeroing out matrix rows corresponding to the discretized variable at those boundary locations. 

There is a *very* simple (maybe too simple) RobinBC in the Squirrel application (a dependency of Zapdos) - [see the link here](https://github.com/arfc/squirrel/blob/d4f54d6b3b7ea653d7adfa941fee4f192018ace2/src/bcs/RobinBC.C). This only takes in a 1D "velocity" value, but could serve as a guide basis for a more complicated custom `IntegratedBC` object that you could create to fit your scenario.

**Reply by hugary1995** [reply]
I may be completely misunderstanding the discussion here. But a ConvectiveHeatFluxBC is the simplest Robin BC based on the canonical definition: https://en.m.wikipedia.org/wiki/Robin_boundary_condition

**Reply by salaudeen-ya** [reply]
I'll take a look at all the suggestions and see what I can make of it.

Thank you all..!

**Comment by salaudeen-ya** [comment]
I'm tried to get a Robin on the right of sub1 and left of sub2, which is
the interface. Not really a direct dirichlet or Neumann on that interface.

**Comment by salaudeen-ya** [comment]
I was wondering if I could take a look at the ConvectiveHeatFluxBC file but
it's registered in Heat Conduction App and not in MOOSE App.

How can I get that please?

Thanks.

**Comment by GiudGiud** [comment]
You can include that module in your application by modifying the makefile. 

If you just want to look it s in modules/heat_conduction/src/bcs

**Comment by salaudeen-ya** [comment]
Thank you.
